File: 1. Creational Design Patterns/Abstract Factory/Go/main.go
package main

import "fmt"

// Button is product A
type Button interface {
	Paint()
}

// Checkbox is product B
type Checkbox interface {
	Paint()
}

// WindowsButton is concrete product A1
type WindowsButton struct{}

func (WindowsButton) Paint() {
	fmt.Println("Render a button in Windows style.")
}

// MacButton is concrete product A2
type MacButton struct{}

func (MacButton) Paint() {
	fmt.Println("Render a button in macOS style.")
}

// WindowsCheckbox is concrete product B1
type WindowsCheckbox struct{}

func (WindowsCheckbox) Paint() {
	fmt.Println("Render a checkbox in Windows style.")
}

// MacCheckbox is concrete product B2
type MacCheckbox struct{}

func (MacCheckbox) Paint() {
	fmt.Println("Render a checkbox in macOS style.")
}

// GUIFactory is the abstract factory
type GUIFactory interface {
	CreateButton() Button
	CreateCheckbox() Checkbox
}

// WindowsFactory is concrete factory 1
type WindowsFactory struct{}

func (WindowsFactory) CreateButton() Button   { return WindowsButton{} }
func (WindowsFactory) CreateCheckbox() Checkbox { return WindowsCheckbox{} }

// MacFactory is concrete factory 2
type MacFactory struct{}

func (MacFactory) CreateButton() Button   { return MacButton{} }
func (MacFactory) CreateCheckbox() Checkbox { return MacCheckbox{} }

// Application is the client
type Application struct {
	button   Button
	checkbox Checkbox
}

func NewApplication(f GUIFactory) *Application {
	return &Application{button: f.CreateButton(), checkbox: f.CreateCheckbox()}
}

func (a Application) Paint() {
	a.button.Paint()
	a.checkbox.Paint()
}

func main() {
	var factory GUIFactory
	factory = WindowsFactory{}
	// Use MacFactory{} for macOS style
	app := NewApplication(factory)
	app.Paint()
}

File: 1. Creational Design Patterns/Abstract Factory/Rust/main.rs
// Abstract Factory Pattern – Rust Example
// Build & run: cargo run --example abstract_factory [windows|mac]

pub trait Button {
    fn paint(&self);
}

pub trait Checkbox {
    fn paint(&self);
}

pub struct WindowsButton;
impl Button for WindowsButton {
    fn paint(&self) {
        println!("Render a button in Windows style.");
    }
}

pub struct MacButton;
impl Button for MacButton {
    fn paint(&self) {
        println!("Render a button in macOS style.");
    }
}

pub struct WindowsCheckbox;
impl Checkbox for WindowsCheckbox {
    fn paint(&self) {
        println!("Render a checkbox in Windows style.");
    }
}

pub struct MacCheckbox;
impl Checkbox for MacCheckbox {
    fn paint(&self) {
        println!("Render a checkbox in macOS style.");
    }
}

pub trait GUIFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
}

pub struct WindowsFactory;
impl GUIFactory for WindowsFactory {
    fn create_button(&self) -> Box<dyn Button> { Box::new(WindowsButton) }
    fn create_checkbox(&self) -> Box<dyn Checkbox> { Box::new(WindowsCheckbox) }
}

pub struct MacFactory;
impl GUIFactory for MacFactory {
    fn create_button(&self) -> Box<dyn Button> { Box::new(MacButton) }
    fn create_checkbox(&self) -> Box<dyn Checkbox> { Box::new(MacCheckbox) }
}

fn paint_app(factory: &dyn GUIFactory) {
    let button = factory.create_button();
    let checkbox = factory.create_checkbox();
    button.paint();
    checkbox.paint();
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && args[1].eq_ignore_ascii_case("mac") {
        paint_app(&MacFactory);
    } else {
        paint_app(&WindowsFactory);
    }
}
